luoxu/auth.py

```diff
@@ -0,0 +1,55 @@
+import secrets
+import hashlib
+import hmac
+import time
+import json
+import base64
+
+from .lib.expiringdict import ExpiringDict
+
+class TokenManager:
+    def __init__(self, ttl):
+        self._token_dict = ExpiringDict(default_ttl=ttl)
+
+    def add_token(self, group_id):
+        self._token_dict.expire()
+        if existing_token := self._token_dict.get(group_id):
+            return existing_token
+        new_token = secrets.token_urlsafe(8)
+        self._token_dict[group_id] = new_token
+        return new_token
+
+    def is_valid(self, group_id, token):
+        self._token_dict.expire()
+        if not (existing_token := self._token_dict.get(group_id)):
+            return False
+        # hmac.compare_digest?
+        return existing_token == token
+
+def Verify_telegram_oauth(bot_token, auth_str):
```

Comment:
```suggestion
def verify_telegram_oauth(bot_token, auth_str):
```

Function names should be lowercased.